### ISSUE TITLE
Feature Request:

### ISSUE THREAD

**Issue by AlpSantoGlobalMomentumLLC** [braden-w/whispering#501]
A great feature would be background noise reduction, and it works very well still if there in the background it's some music or some talk of a not so loud YouTube video. Would be a funny, valuable feature for me.

<html>
<body>
<html><head></head><body><p>Th@t‘s e@sy, here’s the decisive list of battle-tested FOSS you can chain together for real-time speech/noise separation and suppression:</p>
<h2>Snapshot — what works, why it matters</h2>
<p>RNNoise, WebRTC-Audio-Processing, DeepFilterNet 2, Facebook Denoiser, SpeexDSP, VoiceFilter-Lite, Demucs v4, and utility wrappers such as NoiseTorch or the OBS “Noise Suppression” filter all run locally, sub-40 ms on a mid-range CPU. Together they cover (a) classical DSP denoising, (b) RNN/Transformer speech enhancement, and (c) speaker-conditioned voice isolation. Combine two virtual audio cables: one funnels the YouTube tab through a <em>voice-removal</em> stage, the other routes your mic through a <em>voice-preserve</em> stage. Merge both into your conferencing app — result: background music fades, your voice stays razor-clean.</p>
<hr>
<h2>1  Core libraries (drop-in C/C++ or Python)</h2>

Project | Tech | Latency* | Licence | Key point
-- | -- | -- | -- | --
RNNoise | GRU-RNN + light DSP | 10-15 ms | BSD-3 | Gold-standard 48 kHz causal suppressor (GitHub, arXiv)
WebRTC Audio Processing | MMSE + AEC/AGC/NS | 10 ms | 3-clause BSD | Ship-ready, powers every browser tab (Telecom R & D)
SpeexDSP | Frequency-domain Wiener | 10 ms | BSD-like | Ultra-light, great on Raspberry Pi (GitHub)
DeepFilterNet 2 | Dual-stage Conv-T + gating | 20 ms | GPL-3 | Beats RNNoise by ~2 dB PESQ, still realtime (GitHub)
Facebook Denoiser | Conv-TasNet waveform | 6-8 ms | MIT | Causal model, PyTorch, GPU optional (GitHub)
VoiceFilter-Lite | Target-speaker mask (TCN) | 20-25 ms | Apache-2 | Keeps your voice, kills the rest (Google GitHub)
Demucs v4 | Hybrid U-Net + Transformer | 20-40 ms | MIT | Fastest open music/voice separator (GitHub)
latency measured frame-to-frame on 48 kHz, 512-sample hop |   |   |   |  

<hr>
<h2>2  Turn-key desktop wrappers</h2>
<ul>
<li>
<p><strong>NoiseTorch (Linux)</strong> — Creates a virtual mic running RNNoise or DeepFilterNet; one click, works in every PipeWire/PulseAudio app (<a href="https://github.com/noisetorch/NoiseTorch?utm_source=chatgpt.com" title="Real-time microphone noise suppression on Linux. - GitHub">GitHub</a>)</p>
</li>
<li>
<p><strong>OBS “Noise Suppression” filter</strong> — Plugin on top of RNNoise; works for any virtual-cable loopback you add (<a href="https://github.com/werman/noise-suppression-for-voice?utm_source=chatgpt.com" title="Noise suppression plugin based on Xiph's RNNoise - GitHub">GitHub</a>)</p>
</li>
<li>
<p><strong>PipeWire Rust filter &amp; PulseAudio <code inline="">module-echo-cancel</code></strong> — Loadable modules exposing WebRTC-NS or RNNoise to all clients (<a href="https://www.reddit.com/r/rust/comments/142mfml/realtime_noise_suppression_for_pipewire_writen_in/?utm_source=chatgpt.com" title="Real-Time Noise Suppression for PipeWire writen in Rust - Reddit">Reddit</a>, <a href="https://telecom.altanai.com/2022/03/09/aec-echo-cancellation-in-webrtc/?utm_source=chatgpt.com" title="AEC (Echo Cancellation) and AGC (Gain Control) in WebRTC">Telecom R &amp; D</a>)</p>
</li>
<li>
<p><strong>FFmpeg realtime graphs</strong> — <code inline="">anlms</code>, <code inline="">afftdn</code> or <code inline="">arnndn</code> filters for CLI pipelines (<a href="https://ffmpeg.org/ffmpeg-filters.html?utm_source=chatgpt.com" title="FFmpeg Filters Documentation">ffmpeg.org</a>)</p>
</li>
</ul>
<hr>
<h2>3  Minimal wiring to solve your YouTube-chat scenario</h2>
<ol>
<li>
<p><strong>Route audio</strong></p>
<ul>
<li>
<p>Windows/macOS → install two virtual cables (VB-Cable, BlackHole).</p>
</li>
<li>
<p>Linux → create two PipeWire nodes (<code inline="">pactl load-module …</code>).</p>
</li>
</ul>
</li>
<li>
<p><strong>Suppress the YouTube track</strong></p>
<ul>
<li>
<p>Pipe chain: <code inline="">cable-in → Demucs (vocals-only) → RNNoise (residual hiss) → cable-out</code>.</p>
</li>
<li>
<p>Outcome: music/video speech vanishes, ambience kept low.</p>
</li>
</ul>
</li>
<li>
<p><strong>Enhance your microphone</strong></p>
<ul>
<li>
<p>Pipe chain: <code inline="">mic-in → VoiceFilter-Lite (conditioned on your voiceprint) → DeepFilterNet 2 (extra polish) → cable-out</code>.</p>
</li>
</ul>
</li>
<li>
<p><strong>Merge &amp; monitor</strong></p>
<ul>
<li>
<p>In Zoom/Discord set “input” = mic-clean cable, “audio share” = YT-clean cable.</p>
</li>
<li>
<p>OBS users: add two sources and toggle as needed.</p>
</li>
</ul>
</li>
</ol>
<p>Everything here is FOSS, CPU-friendly, and scriptable; you’ll be live in under 15 minutes.</p></body></html>
</body>
</html>Th\@t‘s e\@sy, here’s the decisive list of battle-tested FOSS you can chain together for real-time speech/noise separation and suppression:

## Snapshot — what works, why it matters

RNNoise, WebRTC-Audio-Processing, DeepFilterNet 2, Facebook Denoiser, SpeexDSP, VoiceFilter-Lite, Demucs v4, and utility wrappers such as NoiseTorch or the OBS “Noise Suppression” filter all run locally, sub-40 ms on a mid-range CPU. Together they cover (a) classical DSP denoising, (b) RNN/Transformer speech enhancement, and (c) speaker-conditioned voice isolation. Combine two virtual audio cables: one funnels the YouTube tab through a *voice-removal* stage, the other routes your mic through a *voice-preserve* stage. Merge both into your conferencing app — result: background music fades, your voice stays razor-clean.

---

## 1  Core libraries (drop-in C/C++ or Python)

| Project                                                     | Tech                       | Latency\* | Licence      | Key point                                                        |
| ----------------------------------------------------------- | -------------------------- | --------- | ------------ | ---------------------------------------------------------------- |
| **RNNoise**                                                 | GRU-RNN + light DSP        | 10-15 ms  | BSD-3        | Gold-standard 48 kHz causal suppressor ([[GitHub](https://github.com/xiph/rnnoise?utm_source=chatgpt.com)][1], [[arXiv](https://arxiv.org/abs/1709.08243?utm_source=chatgpt.com)][2]) |
| **WebRTC Audio Processing**                                 | MMSE + AEC/AGC/NS          | 10 ms     | 3-clause BSD | Ship-ready, powers every browser tab ([[Telecom R & D](https://telecom.altanai.com/2022/03/09/aec-echo-cancellation-in-webrtc/?utm_source=chatgpt.com)][3])        |
| **SpeexDSP**                                                | Frequency-domain Wiener    | 10 ms     | BSD-like     | Ultra-light, great on Raspberry Pi ([[GitHub](https://github.com/TeaPoly/speexdsp-ns-python?utm_source=chatgpt.com)][4])                 |
| **DeepFilterNet 2**                                         | Dual-stage Conv-T + gating | 20 ms     | GPL-3        | Beats RNNoise by \~2 dB PESQ, still realtime ([[GitHub](https://github.com/Rikorose/DeepFilterNet?utm_source=chatgpt.com)][5])       |
| **Facebook Denoiser**                                       | Conv-TasNet waveform       | 6-8 ms    | MIT          | Causal model, PyTorch, GPU optional ([[GitHub](https://github.com/facebookresearch/denoiser?utm_source=chatgpt.com)][6])                |
| **VoiceFilter-Lite**                                        | Target-speaker mask (TCN)  | 20-25 ms  | Apache-2     | Keeps *your* voice, kills the rest ([[Google GitHub](https://google.github.io/speaker-id/publications/VoiceFilter-Lite/?utm_source=chatgpt.com)][7])          |
| **Demucs v4**                                               | Hybrid U-Net + Transformer | 20-40 ms  | MIT          | Fastest open music/voice separator ([[GitHub](https://github.com/facebookresearch/demucs?utm_source=chatgpt.com)][8])                 |
| *latency measured frame-to-frame on 48 kHz, 512-sample hop* |                            |           |              |                                                                  |

---

## 2  Turn-key desktop wrappers

* **NoiseTorch (Linux)** — Creates a virtual mic running RNNoise or DeepFilterNet; one click, works in every PipeWire/PulseAudio app ([[GitHub](https://github.com/noisetorch/NoiseTorch?utm_source=chatgpt.com)][9])
* **OBS “Noise Suppression” filter** — Plugin on top of RNNoise; works for any virtual-cable loopback you add ([[GitHub](https://github.com/werman/noise-suppression-for-voice?utm_source=chatgpt.com)][10])
* **PipeWire Rust filter & PulseAudio `module-echo-cancel`** — Loadable modules exposing WebRTC-NS or RNNoise to all clients ([[Reddit](https://www.reddit.com/r/rust/comments/142mfml/realtime_noise_suppression_for_pipewire_writen_in/?utm_source=chatgpt.com)][11], [[Telecom R & D](https://telecom.altanai.com/2022/03/09/aec-echo-cancellation-in-webrtc/?utm_source=chatgpt.com)][3])
* **FFmpeg realtime graphs** — `anlms`, `afftdn` or `arnndn` filters for CLI pipelines ([[ffmpeg.org](https://ffmpeg.org/ffmpeg-filters.html?utm_source=chatgpt.com)][12])

---

## 3  Minimal wiring to solve your YouTube-chat scenario

1. **Route audio**

   * Windows/macOS → install two virtual cables (VB-Cable, BlackHole).
   * Linux → create two PipeWire nodes (`pactl load-module …`).

2. **Suppress the YouTube track**

   * Pipe chain: `cable-in → Demucs (vocals-only) → RNNoise (residual hiss) → cable-out`.
   * Outcome: music/video speech vanishes, ambience kept low.

3. **Enhance your microphone**

   * Pipe chain: `mic-in → VoiceFilter-Lite (conditioned on your voiceprint) → DeepFilterNet 2 (extra polish) → cable-out`.

4. **Merge & monitor**

   * In Zoom/Discord set “input” = mic-clean cable, “audio share” = YT-clean cable.
   * OBS users: add two sources and toggle as needed.

Everything here is FOSS, CPU-friendly, and scriptable; you’ll be live in under 15 minutes.

[1]: https://github.com/xiph/rnnoise?utm_source=chatgpt.com "xiph/rnnoise: Recurrent neural network for audio noise reduction"
[2]: https://arxiv.org/abs/1709.08243?utm_source=chatgpt.com "A Hybrid DSP/Deep Learning Approach to Real-Time Full-Band Speech Enhancement"
[3]: https://telecom.altanai.com/2022/03/09/aec-echo-cancellation-in-webrtc/?utm_source=chatgpt.com "AEC (Echo Cancellation) and AGC (Gain Control) in WebRTC"
[4]: https://github.com/TeaPoly/speexdsp-ns-python?utm_source=chatgpt.com "Python bindings of speexdsp noise suppression library - GitHub"
[5]: https://github.com/Rikorose/DeepFilterNet?utm_source=chatgpt.com "Rikorose/DeepFilterNet: Noise supression using deep filtering"
[6]: https://github.com/facebookresearch/denoiser?utm_source=chatgpt.com "Real Time Speech Enhancement in the Waveform Domain ... - GitHub"
[7]: https://google.github.io/speaker-id/publications/VoiceFilter-Lite/?utm_source=chatgpt.com "VoiceFilter-Lite: Streaming Targeted Voice Separation for On-Device ..."
[8]: https://github.com/facebookresearch/demucs?utm_source=chatgpt.com "facebookresearch/demucs: Code for the paper Hybrid ... - GitHub"
[9]: https://github.com/noisetorch/NoiseTorch?utm_source=chatgpt.com "Real-time microphone noise suppression on Linux. - GitHub"
[10]: https://github.com/werman/noise-suppression-for-voice?utm_source=chatgpt.com "Noise suppression plugin based on Xiph's RNNoise - GitHub"
[11]: https://www.reddit.com/r/rust/comments/142mfml/realtime_noise_suppression_for_pipewire_writen_in/?utm_source=chatgpt.com "Real-Time Noise Suppression for PipeWire writen in Rust - Reddit"
[12]: https://ffmpeg.org/ffmpeg-filters.html?utm_source=chatgpt.com "FFmpeg Filters Documentation"
